geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs

```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+
+        case ohlcvInfo of
+          Left e   → return . Left $ SourceUnavailable (displayException e) "Taptools"
+          Right [] → return . Left $ SourceUnavailable "Empty OHLCV." "Taptools"
+          Right (ttOHLCV : _) → do
+            let price         = close ttOHLCV
+                precisionDiff = 10 ** fromIntegral (mmtPrecision mmtLovelace - precision)
+                adjustedPrice = price * precisionDiff
+            return . Right . Price . toRational $ adjustedPrice
+        where
+          withoutQuotes ∷ String → String
+          withoutQuotes s = case s of
+            ('"':xs) | not (null xs) && last xs == '"' → init xs
+            _                                          → s
+
+  _         → throwIO $ userError "Only Mainnet is supported."
+
+buildGetQuota _ (MockPPC MockConfig {..}) = GetQuota $ \_ → do
+  let price = mokcPrice
+  
+  offset ← case mokcOffsetPath of
+    Nothing   → pure 0
+    Just path → readOffset path `catch` handleOffsetReadError
+
+  let adjustedPrice = price + offset
+
+  mockAvailable ← case mokcAvailablePath of
+    Nothing   → pure True
+    Just path → readBool path `catch` handleBoolReadError
+
+  if mockAvailable
+    then return . Right . Price . toRational $ adjustedPrice
+    else return . Left $ SourceUnavailable "Mock prices provider exception." "Mock Prices Provider"
+
+buildGetQuotas ∷ PricesProviders → [GetQuota]
+buildGetQuotas PP {pricesAggregatorPP = PPA {..}} = buildGetQuota ppaCommonCfg <$> ppaPricesCluster
+
+buildWeights ∷ PricesProviders → [Int]
+buildWeights PP {pricesAggregatorPP = PPA {..}} = pccPricesProvidersWeights ppaCommonCfg
+
+
+-------------------------------------------------------------------------------
+-- Price "estimate" from providers
+-------------------------------------------------------------------------------
+
+priceEstimate ∷ PricesProviders → MMTokenPair → IO PriceIndicator
+priceEstimate pp mmtp = do
+  let pproviders = buildGetQuotas pp
+      weights    = buildWeights pp
+
+  eps ← mapM (\prov → getQuota prov mmtp) pproviders
+  
+  let weightedPrices = catMaybes $ zipWith matchRight weights eps
+
+  let lps                   = lefts eps
+      hasSourceFailed       = not . null $ lps
+      unavailableExceptions = map (\(SourceUnavailable e _) → e) lps
+      unavailablePProviders = map (\(SourceUnavailable _ pprovider) → pprovider) lps
+   
+  case weightedPrices of
+    []  → return PriceUnavailable
+    wps → do
+      let p    = Price . weightedMean . NE.fromList $ wps
+          rsd  = relStdDev . NE.fromList $ fromRational . snd <$> wps
+          cCfg = ppaCommonCfg . pricesAggregatorPP $ pp
+          ths1 = pccPriceDiffThreshold1 cCfg
+          ths2 = pccPriceDiffThreshold2 cCfg
+
+          priceAnalyzed
+            | rsd > ths2      = PriceMismatch2
+            | rsd > ths1      = PriceMismatch1
+            | hasSourceFailed = PriceSourceFail unavailableExceptions unavailablePProviders p
```

Comment:
`unavailableExceptions` could be abstracted as a datatype listing all available providers exception, like in `GeniusYield.TxBuilder.Errors` module inside Atlas.